src/core/SampleBuffer.cpp

```diff
@@ -293,123 +217,112 @@ void SampleBuffer::update( bool _keep_settings )
-		for( f_cnt_t frame = 0; frame < _frames;
-						++frame )
-		{
-			m_data[frame][0] = _ibuf[idx+0] * fac;
-			m_data[frame][1] = _ibuf[idx+ch] * fac;
-			idx -= _channels;
-		}
-	}
-	else
+	int idx = 0;
+	for( f_cnt_t frame = 0; frame < _frames;
+					++frame )
 	{
-		int idx = 0;
-		for( f_cnt_t frame = 0; frame < _frames;
-						++frame )
-		{
-			m_data[frame][0] = _ibuf[idx+0] * fac;
-			m_data[frame][1] = _ibuf[idx+ch] * fac;
-			idx += _channels;
-		}
+		vector[frame][0] = _ibuf[idx+0] * fac;
+		vector[frame][1] = _ibuf[idx+ch] * fac;
+		idx += _channels;
 	}
 
 	delete[] _ibuf;
-}
 
-void SampleBuffer::directFloatWrite ( sample_t * & _fbuf, f_cnt_t _frames, int _channels)
-
-{
-
-	m_data = MM_ALLOC( sampleFrame, _frames );
-	const int ch = ( _channels > 1 ) ? 1 : 0;
-
-	// if reversing is on, we also reverse when
-	// scaling
-	if( m_reversed )
-	{
-		int idx = ( _frames - 1 ) * _channels;
-		for( f_cnt_t frame = 0; frame < _frames;
-						++frame )
-		{
-			m_data[frame][0] = _fbuf[idx+0];
-			m_data[frame][1] = _fbuf[idx+ch];
-			idx -= _channels;
-		}
-	}
-	else
-	{
-		int idx = 0;
-		for( f_cnt_t frame = 0; frame < _frames;
-						++frame )
-		{
-			m_data[frame][0] = _fbuf[idx+0];
-			m_data[frame][1] = _fbuf[idx+ch];
-			idx += _channels;
-		}
-	}
-
-	delete[] _fbuf;
+	return vector;
 }
 
 
 void SampleBuffer::normalizeSampleRate( const sample_rate_t _src_sr,
 							bool _keep_settings )
 {
 	// do samplerate-conversion to our default-samplerate
-	if( _src_sr != Engine::mixer()->baseSampleRate() )
+	if( _src_sr != sampleRate ())
 	{
-		SampleBuffer * resampled = resample( _src_sr,
-					Engine::mixer()->baseSampleRate() );
-		MM_FREE( m_data );
-		m_frames = resampled->frames();
-		m_data = MM_ALLOC( sampleFrame, m_frames );
-		memcpy( m_data, resampled->data(), m_frames *
-							sizeof( sampleFrame ) );
-		delete resampled;
+		m_data = resampleData (m_data, _src_sr, sampleRate ());
 	}
 
 	if( _keep_settings == false )
 	{
 		// update frame-variables
 		m_loopStartFrame = m_startFrame = 0;
-		m_loopEndFrame = m_endFrame = m_frames;
+		m_loopEndFrame = m_endFrame = frames ();
 	}
 }
 
 
+SampleBuffer::DataVector
+SampleBuffer::resampleData (const DataVector &inputData, sample_rate_t inputSampleRate,
+							sample_rate_t requiredSampleRate)
+{
+	const f_cnt_t dst_frames = static_cast<f_cnt_t>( inputData.size ()/
+					(float) inputSampleRate * (float) requiredSampleRate );
+	DataVector outputData(dst_frames);
 
+	// yeah, libsamplerate, let's rock with sinc-interpolation!
+	int error;
+	SRC_STATE * state;
+	if( ( state = src_new( SRC_SINC_MEDIUM_QUALITY,
+					DEFAULT_CHANNELS, &error ) ) != NULL )
+	{
+		SRC_DATA src_data;
+		src_data.end_of_input = 1;
+		src_data.data_in = libSampleRateSrc(inputData.data ())->data ();
+		src_data.data_out = outputData.data ()->data();
+		src_data.input_frames = inputData.size ();
+		src_data.output_frames = dst_frames;
+		src_data.src_ratio = (double) inputSampleRate / requiredSampleRate;
```

Comment:
This line leads to weird results when resampling. You should swap denominator and numerator.